tests/Highlighter.py

```diff
@@ -0,0 +1,303 @@
+
+            # search for the transition that we will use
+            current_match = None
+            NEXT = {"rule":{},"match":None}
+            FIND = False
+            next_pos = len(code) + 42
+            for rule in rules[current_state]:
+
+                regex_compile = re.compile(rule['regex'], re.MULTILINE)
+                match = regex_compile.search(code, current_position)
+
+                if match:
+                    if match.start() < next_pos :
+                        next_pos = match.start()
+                        NEXT["rule"] = rule
+                        NEXT["match"] = match
+                        FIND = True
+
+
+            if FIND :
+
+                # Application of coloring on the code
+                current_rule,current_match = NEXT["rule"],NEXT["match"]
+
+                if "token" not in current_rule:
+                    raise ValueError("We need a token in all rules !")
+        
+                if type(current_rule['token']) == str :
+                    current_rule['token'] = [current_rule['token']]
+        
+                if re.compile(current_rule['regex'], re.MULTILINE).groups == 0:
+                    tok = current_rule['token'][0]
+                    start = current_match.start()
+                    length = current_match.end() - current_match.start()
+                    output = output[:start] + TOKEN_CODE[tok] * length + output[start + length:]
+
+                else:
+                    pos = current_match.start()
+                    for i,submatch in enumerate(current_match.groups()):    
+                        tok = current_rule['token'][i%len(current_rule['token'])]        
+                        output = output[:pos] + TOKEN_CODE[tok] * len(submatch) + output[pos + len(submatch):]
+                        pos += len(submatch)
+
+                current_position = current_match.end()
+
+                if 'next' in current_rule:
+                    current_state = current_rule['next']
+
+                if current_position ==len(code):
+                    flag = True
+
+
+            else:
+                flag = True
+
+        output = output.replace(" ","T")
+        return output,current_state
+
+
+    def apply_rules(self,rules,code):
+        outputs = []
+        token = "start"
+        for line in code.split("\n"):
+            output,token = self.apply_rules_line(rules,line,token)
+            print("next",token)
+            outputs.append(output)
+        return "\n".join(outputs)
+
+
+
+    # This function allows to check 2 syntactic colorations one desired
+    # and the other obtained, and manages some special cases
+
+    # Returns a boolean and a value:
+    # - If the 2 highlights are consistent, the boolean will be true and the value unused
+    # - if there is an error in the highlight, the boolean will be false and
+    #   the value will be the number of the 1st character where there is an inconsistency
+    def check(self,result,expected):
```

Comment:
Just use `self.assertEquals()` instead of writing your own matcher function.

When you are comparing expected to actual, you want as much information as possible, and the built-in functions do a pretty good job of printing a diff already. No need to rebuild that.

Reply:
But then we could `in`? Something like: `self.assertIn(result, ["KKK TTTTTTTTT", "KKKKTTTTTTTTT", "KKKTTTTTTTTTT"])` so we test if we get one of the 3 valid answers?

Reply:
Yes, but that means generating a lot of str :

In this example: 
```
if throw is earthworm points = points + 5 else points = points + throw
KK TTTTT KK TTTTTTTTT TTTTTT K TTTTTT K N KKKK TTTTTT K TTTTTT K TTTTT
```
that would be 6^15 possible strings

The solution could be to pass to assert some function of comparison between characters, but the only way I found to do that is to create a new object.

Reply:
ok, I see that might yield too many options, but then we could wrap our own equality function into the assert? Something like:

```
s = "ask something"

result = remove_space_highlighting(highlight(s), 4). (so we'd get "KKKTTTTTTTTT" for "ask something")

self.assert(result,  "KKKTTTTTTTTT")

def remove_space_highlighting(s, space):
   # removes the highlighting symbol at the location of the space
  return s[:space] + s[space+1:]

```

Hope that makes sense!

Reply:
We can't just delete the undefined characters either, because my function doesn't allow coloring cases for undefined characters.

In particular, an indefinite character can be anything, except a character highlighted in pink as in the following case :

```
if a is 2 else _ "Done!"
KK T KK N KKKK I SSSSSSS
              | |
              1 2
```

Spaces 1 and 2 can be colored with the style "command" or "string" or "text", because it will not be seen, but they should not be colored with the style "error" (which is the one of the pink highlighting)

Reply:
ah ok but then we can make a function that checks whether these spaces are "not error"?

We just really don't want our own assertion function, we want something (does not matter how you divide it precisely) wrapped inside the regular assert.

Reply:
I tested a new solution, it splits the check in 2 different assert (In and Equal) but it can be enough

Reply:
> For example, if we have the code :
>
> ```
> "ask something"
> "KKK TTTTTTTTT" (1)
> "KKKKTTTTTTTTT" (2)
> "KKKTTTTTTTTTT" (3)
> ```
>
> We want this coloration (1) But (2) or (3) are valid too

Well that's a bit odd, isn't it? But also, this is nothing a bit of *canonicalization* wouldn't fix.

Using the following rule:

**Whitespace will never be assigned a color**

We could say the one and only colorization of this string is `"KKK TTTTTTTTT"`, and the other 2 variants are not allowed. This seems reasonable, as a human wouldn't be able to perceive the difference anyway.

The canonicalization can be applied in one of 2 ways:

* Either just before testing (requires both the original code string and the colorized string, to find the whitespace locations to strip out)
* Or already during the highlighter simulation. The `AceHighlighter` code could take into account whether or not the input string contains whitespace and remove the highlight tokens there.

The canonicalization function could look something like this:

```py
# Array of booleans
is_whitespace = [c == ' ' for c in code]

# Do a transform on the colorization, retaining the characters only for non-whitespace
# (whitespace characters are replaced with a space)
canonical_colors = ''.join(c if not is_ws else ' ' for c, is_ws in zip(colorization, is_whitespace))
```

Reply:
Haha thanks @rix0rrrm that is a more extensive explanation of what I suggested above :)

Reply:
I understand better about channeling, but it only works for spaces.

However, trying to adapt the method, there is another problem concerning uncertain characters. 

Indeed, there are the space characters, which can be colored as "keyword", "number" or "string" it will not matter, because it will not be seen on the space, and that's true.
But there are other cases where you want the coloring to be uncertain, and not on spaces!

For example, in this case :
```
print _ the dishes are done by _
KKKKK I                        I
```
because the code is incomplete, the coloring is uncertain.

I don't think that fixing the coloring in this case is a good idea, and that's also why I've taken this view of uncertain coloring.

If you think it's not a good idea to allow tests to be written like this, and the new way of checking tests is not appropriate, then I could delete the spaces afterwards.

Reply:
> because the code is incomplete, the coloring is uncertain.

Just to clarify, we should have a special coloring for incomplete code (very pink):

<img width="407" alt="image" src="https://user-images.githubusercontent.com/1003685/170865596-6c591ff8-1282-4caa-99b8-ad0d8ed94528.png">

So, I guess you mean here is the part "the dishes are done by" Yes, we do not know the coloring, but then we can just not have these types of tests? Because it makes no sense to test things we don't have the correct answer to.

I would propose not to test these uncertainties at all, also then just proceed with the spaces as we planned earlier. Just test `print _` and verify it is keyword followed by incomplete.